### DISCUSSION TITLE
Nuxt Fonts

### DISCUSSION THREAD

**Discussion by danielroe** [nuxt/nuxt#22014] [RFCs]
# Nuxt Fonts RFC

## Summary

The `nuxt/fonts` integration is a new addition to the Nuxt framework that allows developers to easily configure fonts in their Nuxt apps (whether system or web fonts). This is relevant directly for end users as well as modules (like `@nuxtjs/tailwindcss` or `@nuxtjs/vuetify`).

## Implementation

`nuxt/fonts` integrates with the Nuxt Assets RFC to automatically download fonts and font CSS as required as part of the build process. It also automatically generates CSS to reduce layout shift and adds prefetch/preload/dns-preconnect hints as required.

### Providers

Different providers are available to handle the loading of web fonts. Out-of-the-box, `nuxt/fonts` will ship with `local` and `google` providers, with PRs accepted to add more to core, though modules will also be able to register their own providers.

Providers will be passed font family definitions specified in the configuration and will be able to handle registering assets using Nuxt Assets (see RFC) as well as inject CSS globally or per-use.

**Inspirations**: <https://unocss.dev/presets/web-fonts#providers>

### API

In most cases, `nuxt/fonts` will be able to detect your font usage and optimise your fonts automatically throughout your app without any explicit configuration.

However, it will be possible to specify fonts and/or default providers to be used in your app using a new top-level `fonts` configuration option within your `nuxt.config`, specifying the available fonts within your app.

```ts
export default defineNuxtConfig({
  fonts: {
    provider: 'google', // sets default provider
    families: [
      {
        name: 'Roboto', // the 'canonical' name of the font used to look it up in a provider database
        as: 'custom-roboto-family', // allow registering a font family with a different name
        provider: 'local', // you can override the provider on a per-family basis
        // provider specific options can be provided
        src: '~/public/roboto.woff2', // you can specify a source within your project
        // specific configuration will be used to generate `@font-face` definitions
        subsets: ['latin', 'greek'],
        display: 'swap', // or 'block'
        weight: ['400', '700'],
        style: ['normal', 'italic'],
        // and produce CSS overrides to reduce layout shift (using fontaine)
        fallbacks: ['Arial'],
      }
    ]
  }
})
```

**Inspirations**: <https://nextjs.org/docs/app/building-your-application/optimizing/fonts>

**Comment by therealokoro** [comment]
Awesome!. I am particularly excited about this one.

I hope variable fonts would work out of the box. I also think fontsource should be part of the providers supported out of the box

**Comment by alvarosabu** [comment]
Yes and Yes

**Comment by MuhammadM1998** [comment]
To confirm my understanding that this is will be out of the box with no modules installation needed? I just need to set the `fonts` in `nuxt.config` and i'm ready to go?

**Reply by therealokoro** [reply]
Pretty much so 🤷🏾 I guess

**Comment by kissu** [comment]
Could we have something similar to the `text` option (as in [Google Fonts](https://developers.google.com/fonts/docs/getting_started#optimizing_your_font_requests)) to have only specific letters for a given font for further optimization?

**Comment by kissu** [comment]
Not sure about all the benefits of base64 but this one may be useful to some people? 🤔
https://google-fonts.nuxtjs.org/downloading#base64

**Comment by AndreyYolkin** [comment]
Nice! looks like https://github.com/nuxt-modules/fontaine can also benefit from this

**Comment by jonrobinson** [comment]
This is just what I need for a project I'm working on!

**Comment by ThimoDEV** [comment]
Really looking forward to see this integration into nuxt core. I have been using the NextJS implementation of this and its such a joy to work with. Being able to load fonts dynamically or locally, specify a variable name, and then import it not having to think about it anymore

**Comment by ricardogobbosouza** [comment]
I started working on the module
https://github.com/ricardogobbosouza/nuxt-fonts

**Comment by baixiaoyu2997** [comment]
does it support work with nuxt/i18n ?

**Comment by TheAlexLichter** [comment]
Linking the now existing package - https://github.com/nuxt/fonts

**Comment by tubstrr** [comment]
Question about [Adobe Font licensing](https://helpx.adobe.com/fonts/using/font-licensing.html#web-customize) (and potentially other providers)

On Adobe Font licensing they answer this question: 
> ## Can I host the web font files on my own server?
> No. Adobe doesn't offer the ability to locally host fonts. Our web font hosting delivers fonts from a globally-distributed content delivery network ([CDN](https://en.wikipedia.org/wiki/Content_delivery_network)), which performs much better and is more scalable than using local hosting for static assets.

On the [@nuxtjs/fonts](https://github.com/nuxt/fonts?tab=readme-ov-file#how-it-works) README.md it says:
> 3. **Proxies and caches font requests.** Rather than using the original source URLs (to remote servers), we generate rewrites under the `/_fonts` subpath. When accessed by your browser, we download the font from the remote server and cache it locally.

Isn't this process a violation of Adobe's licensing?

**Reply by kissu** [reply]
I think it's more of a difference of having it as a single file offline or using an authenticated CDN. In the end I think that everything is alright. Also using a CDN nowadays is bad for a lot of reasons. Having it locally installed from the CDN is all good in terms of licensing and a good practice IMO.

If that wasn't the case the performance would have a big hit and be less reliable. 
Don't Adobe have a tool to see if you're doing things right?

**Comment by rnambs** [comment]
how do i set a font globally? im using nuxt3 and i'm using a font i downloaded. i added it to nuxt.config.js and modified my tailwind config file, but i still have to apply it manually in a css class on each file for it to take effect.

**Reply by kissu** [reply]
We would need some technical details to help you further.
You can start with this: https://stackoverflow.com/a/74603777/8816585

**Reply by AloisSeckar** [reply]
In your `app.vue` add into `<script setup>`:

```
useHead({
  bodyAttrs: {
    class: '<your-tailwind-font-class>'
  }
})
```

This will apply the class on `<body>` tag level making it "global".

**Reply by rnambs** [reply]
Thank you. Turned out to be that issue and a corrupted font file.

**Comment by rw-simon** [comment]
Hello

I don't seem to be able to load OpenType features (alternate glyphs).
Specifically I want to use the font DM Sans from google (which works fine), but use alternate glyphs for a and y. 

I tried the font-feature-settings, but to no avail. Also font-variant-alternates and similar css commands i found online did not work.
If i use the installed font directly from the client (so without nuxt-font), it does work. How do i need to import the font to have the opentype variants available?

Thanks
Simon

**Reply by TheAlexLichter** [reply]
Best would be to raise an issue in https://github.com/nuxt/fonts with a reproduction 👍🏻
